kjdoc/tour.md

```diff
@@ -538,13 +538,440 @@ This section describes KJ APIs that control process execution and low-level inte
+
+It is especially important to note that once a promise has been canceled, then any references that were received along with the promise may no longer be valid. For example, consider this function:
+
+```
+kj::Promise<void> write(kj::ArrayPtr<kj::byte> data);
+```
+
+The function receives a pointer to some data owned elsewhere. By KJ convention, the caller must ensure this pointer remains valid until the promise completes _or is canceled_. If the caller decides it needs to free the data early, it may do so as long as it cancels the promise first. This property is important as otherwise it becomes impossible to reason about ownership in complex systems.
+
+This means that the implementation of `write()` must immediately stop using `data` as soon as cancellation occurs. For example, if `data` has been placed in some sort of queue where some other concurrent task takes items from the queue to write them, then it must be ensured that `data` will be removed from that queue upon cancellation. This "queued writes" pattern has historically been a frequent source of bugs in KJ code, to the point where experienced KJ developers now become immediately suspicious of such queuing. The `kj::AsyncOutputStream` interface explicitly prohibits overlapping calls to `write()` specifically so that the implementation need not worry about maintaining queues.
+
+### Promise-Fulfiller Pairs and Adapted Promises
+
+Sometimes, it's difficult to express asynchronous control flow as a simple chain of continuations. For example, imagine a producer-consumer queue, where producers and consumers are executing concurrently on the same event loop. The consumer doesn't directly call the producer, nor vice versa, but the consumer would like to wait for the producer to produce an item for consumption.
+
+For these situations, you may use a `Promise`-`Fulfiller` pair.
+
+```c++
+kj::PromiseFulfillerPair<int> paf = kj::newPromiseAndFulfiller<int>();
+
+// Consumer waits for the promise.
+paf.promise.then([](int i) { ... });
+
+// Producer calls the fulfiller to fulfill he promise.
+paf.fulfiller->fulfill(123);
+
+// Producer can also reject the promise.
+paf.fulfiller->reject(KJ_EXCEPTION(FAILED, "something went wrong"));
+```
+
+**WARNING! DANGER!** When using promise-fulfiller pairs, it is very easy to forget about both exception propagation and, more importantly, cancellation-safety.
+
+* **Exception-safety:** If your code stops early due to an exception, it may forget to invoke the fulfiller. Upon destroying the fulfiller, the consumer end will receive a generic, unhelpful exception, merely saying that the fulfiller was destroyed unfulfilled. To aid in debugging, you should make sure to catch exceptions and call `reject()` to propagate them.
```

Comment:
```suggestion
* **Exception-safety:** If your code stops early due to an exception, it may forget to invoke the fulfiller. Upon destroying the fulfiller, the consumer end will receive a generic, unhelpful exception, merely saying that the fulfiller was destroyed unfulfilled. To aid in debugging, you should make sure to catch exceptions and call `fulfiller->reject()` to propagate them.
```